cylc/flow/prerequisite.py

```diff
@@ -79,6 +75,22 @@ def __init__(self, point, start_point=None):
         # * `False` (prerequisite unsatisfied).
         self._all_satisfied = None
 
+    def __eq__(self, other):
+        """Test equality of prerequisites. Ignore derived attributes."""
+        return self.hash() == other.hash()
+
+    def instantaneous_hash(self):
+        """Generate a hash of this prerequisite in its current state.
+
+        Note not using `__hash__` because Prerequisite objects are mutable.
+        """
+        return hash((
+            self.point,
+            self.conditional_expression,
+            tuple(self.satisfied.keys()),
+            tuple(self.satisfied.values()),
```

Comment:
Is `self.satisfied` actually relevant to `TaskState._add_prerequisites()`? And is `self.target_point_strings` not relevant?

Reply:
`self.satisfied` is poorly named, but it's a dict of all prerequisites, so it's really the most important thing (the boolean satisfied/not-satisfied bit is in there, but incidental to this PR).

`self.target_point_strings` just repeats info held in `self.satisfied` so it's not needed in the hash.

Reply:
So I think we only need `satisfied.keys()`.

Reply:
Yes actually.

Reply:
(Signing off for now - over to you guys 👋 )

Reply:
Confirmed that `target_point_strings` is the same as `{message[0] for message in self.satisfied}`.

We don't need to store both so remove `target_point_strings` here: https://github.com/cylc/cylc-flow/pull/5496

Reply:
```suggestion
            tuple(self.satisfied.keys()),
```